### PR TITLE
Enhancement: Deprecating use of langspec

### PR DESCRIPTION
**Pull request by ahangsu**

### CODE REVIEW

File: logic/logic.go
Comment (michaeldiamant) [inline logic/logic.go:35]: @ahangsu Minor:  Based on my read of https://github.com/golang/go/wiki/Deprecated, we ought to tweak deprecation format here.

> To signal that an identifier should not be used, add a paragraph to its doc comment that begins with `Deprecated:` followed by some information about the deprecation, and a recommendation on what to use instead, if applicable.

Comment (ahangsu) [reply logic/logic.go:35]: I see, that makes sense, we can extend the deprecation note from java one to the golang `// Deprecated: ...`.

File: .test-env
Comment (ahangsu) [inline .test-env:4]: after algorand-sdk-testing is merged, do:
```suggestion
SDK_TESTING_BRANCH="master"
```

File: logic/logic.go
Comment (michaeldiamant) [inline logic/logic.go:67]: @ahangsu Let me know what you think of the following proposal?

I propose marking `SanityCheckProgram` (and analogous methods in other SDKs) private.

Since the check is _only_ a set of heuristics, my thinking is that we should _not_ encourage API users to rely on the check.

File: logic/logic.go
Comment (michaeldiamant) [inline logic/logic.go:86]: @ahangsu I appreciate the added context here.  I propose these tweaks:
* Avoid using a date - it's hard to correlate a date with a released version.  Either prefer version where deprecation begins (e.g. v1.20.0 for this repo as of writing) or leaving without specification.
* Focus message on langspec.json inaccuracies rather than maintainenance concerns.

To minimize multiple iterations, I also suggest aligning all reviewers on phrasing.  I'm happy to speak live if it'll help us reach a conclusion faster.

```suggestion
// Deprecated: Validation relies on metadata (`langspec.json`) that does not accurately represent opcode behavior across program versions.
// The behavior of `ReadProgram` relies on `langspec.json`. Thus, this method is being deprecated.
```

File: logic/logic.go
Comment (ahangsu) [reply logic/logic.go:67]: good point, I made it private in js-algorand-sdk, but LogicsigSignature in `crypto` package cannot get access to this heuristic method. Is there a preferred way to make this private, while still can be accessed in other package?

Comment (ahangsu) [reply logic/logic.go:66]: I made the Js/Py-sdk's `sanityCheckProgram` private or protected, but in Java/Go, it does not seem straightforward to both hide this from user and export it internally

Comment (michaeldiamant) [reply logic/logic.go:66]: @ahangsu Here's my thinking:

For Java, provided we agree to mark `sanityCheckProgram` private, then I suggest:
* Move definition of `sanityCheckProgram` into `LogicsigSignature`.
* Mark `public class Logic` as deprecated because it'll contain no non-deprecated functionality. 
  * Implies we can remove deprecations on specific fields + methods because the class is deprecated.

For Go, I offer the same suggestion + caveats with `SanityCheckProgram` being moved into `crypto.go`.

Comment (ahangsu) [reply logic/logic.go:66]: that sounds reasonable to me, working on it now.

Comment (jasonpaulos) [reply logic/logic.go:66]: I've only done a cursory review so far, but I'll leave my thoughts here.

It seems like we cannot avoid changing the behavior of the SDKs in some way. Indeed, that's the goal of this PR. That being the case, I'd opt to change the existing `CheckProgram` function to become `SanityCheckProgram`. That way, we do not export any new functions and we can keep the implementation in this package.

Comment (ahangsu) [reply logic/logic.go]: for function exporting issue, if we move heuristic sanity check function into `LogicsigSignature`, and make them private methods, then we are just simply adding an internal method, rather than something exported. We are deprecating `CheckProgram` anyways, I think...

Comment (jasonpaulos) [reply logic/logic.go]: Ok, I will not push further. The latest changes make sense

File: crypto/crypto.go
Comment (jasonpaulos) [inline crypto/crypto.go:646]: There is an unfortunate problem. This function is deprecated and has been replaced by `MakeLogicSigAccountEscrow`, `MakeLogicSigAccountDelegated`, and `MakeLogicSigAccountDelegatedMsig`.

The good news is that `MakeLogicSigAccountDelegated` and `MakeLogicSigAccountDelegatedMsig` call into `MakeLogicSig`, so this change works for them. The bad news is that `MakeLogicSigAccountEscrow` does not call into `MakeLogicSig`, and even if we modified it to do so, it does not currently return an error response.

Comment (ahangsu) [reply crypto/crypto.go:646]: ugh, my IDE does not show that it has been deprecated, I should probably mark it deprecated as golang syntax.

I think I should be a nice dude to add the check in `MakeLogicSigAccountEscrow `, so the basic check should be reflected...

Comment (ahangsu) [reply crypto/crypto.go:646]: never mind, I just realized `MakeLogicSigAccountEscrow` returns only `LogicSigAccount`, *without error*. That is unfortunate.

Comment (jasonpaulos) [reply crypto/crypto.go:646]: Right, at the moment I think the only solution is to deprecate `MakeLogicSigAccountEscrow` and create a new function that will return an error too.

Comment (ahangsu) [reply crypto/crypto.go:646]: deprecation change reflected in [f8dab81](https://github.com/algorand/go-algorand-sdk/pull/366/commits/f8dab818abe3564d6a577a18f6ea926357cebf62), let me know if you are good with new method name

File: crypto/account.go
Comment (jasonpaulos) [inline crypto/account.go:206]: nit: it would be a little nicer to call `MakeLogicSig` here to unify all the code paths.

File: crypto/crypto.go
Comment (jasonpaulos) [reply crypto/crypto.go:647]: Thanks, new name looks good

File: crypto/account.go
Comment (ahangsu) [reply crypto/account.go:206]: yes, that makes sense

File: test/steps_test.go
Comment (jasonpaulos) [inline test/steps_test.go:2655]: nit: is this printf meant to be here?

**Review by jasonpaulos — COMMENTED** [review]
This PR looks good to me, with the exception of one comment below.

I will come back and approve once the algorand-sdk-testing branch has been updated and that comment is resolved.

File: test/steps_test.go
Comment (ahangsu) [reply test/steps_test.go]: ugh my bad, fat finger

**Review by michaeldiamant — APPROVED** [review]

**Review by jasonpaulos — APPROVED** [review]

**Pull request merged**
